src/net.cpp

```diff
@@ -141,22 +141,23 @@ bool GetLocal(CService& addr, const CNetAddr *paddrPeer)
     return nBestScore >= 0;
 }
 
-//! Convert the pnSeed6 array into usable address objects.
-static std::vector<CAddress> convertSeed6(const std::vector<SeedSpec6> &vSeedsIn)
+//! Convert the serialized seeds into usable address objects.
+static std::vector<CAddress> convertSeeds(const std::vector<uint8_t> &vSeedsIn)
 {
     // It'll only connect to one or two seed nodes because once it connects,
     // it'll get a pile of addresses with newer timestamps.
     // Seed nodes are given a random 'last seen time' of between one and two
     // weeks ago.
     const int64_t nOneWeek = 7*24*60*60;
     std::vector<CAddress> vSeedsOut;
-    vSeedsOut.reserve(vSeedsIn.size());
     FastRandomContext rng;
-    for (const auto& seed_in : vSeedsIn) {
-        struct in6_addr ip;
-        memcpy(&ip, seed_in.addr, sizeof(ip));
-        CAddress addr(CService(ip, seed_in.port), GetDesirableServiceFlags(NODE_NONE));
+    CDataStream s(vSeedsIn, SER_NETWORK, PROTOCOL_VERSION | ADDRV2_FORMAT);
```

Comment:
With a `VectorReader` here you'd avoid copying the encoded bytes.

Reply:
Yea—a lot of optimizations could be made here; it's a byte array in rodata, we could return it as `span<uint8_t>`, and parse directly from there, making it zero-copy. That said, this happens once in the lifetime of a node (ideally) and it's not that much data, so i'm not sure.
But if it's just a matter of swapping the type used i'm in :smile:

Reply:
I think I'm not sure how to use it. At least the obvious change gives a lot of compiler errors:
```diff
diff --git a/src/net.cpp b/src/net.cpp
index 52e35b56077bf6562202461f73a93e03097c91f9..507d61df0cd541e0696ead1c22cc582380bb1521 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -151,8 +151,8 @@ static std::vector<CAddress> convertSeeds(const std::vector<uint8_t> &vSeedsIn)
     const int64_t nOneWeek = 7*24*60*60;
     std::vector<CAddress> vSeedsOut;
     FastRandomContext rng;
-    CDataStream s(vSeedsIn, SER_NETWORK, PROTOCOL_VERSION | ADDRV2_FORMAT);
-    while (!s.eof()) {
+    VectorReader s(SER_NETWORK, PROTOCOL_VERSION | ADDRV2_FORMAT, vSeedsIn, 0);
+    while (!s.empty()) {
         CService endpoint;
         s >> endpoint;
         CAddress addr(endpoint, GetDesirableServiceFlags(NODE_NONE));
```

```
/…/bitcoin/src/netaddress.h:430:15: error: invalid operands to binary expression ('VectorReader' and 'Wrapper<CompactSizeFormatter<true>, unsigned long &>')                            
            s >> COMPACTSIZE(address_size);
            ~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                          
```

Reply:
Ok, never mind then. We can figure out how to make this work in a follow-up (there are probably several other cases in the codebase where this is done).